modules/core/src/test/java/org/apache/ignite/internal/causality/VersionedValueTest.java

```diff
@@ -260,189 +198,47 @@ public void testAutocompleteFuture() throws OutdatedTokenException {
-
-        assertFalse(fut.isDone());
-
-        longVersionedValue.update(0, (previous, e) -> {
-            assertNull(previous);
-
-            return completedFuture(TEST_VALUE);
-        });
-
-        assertFalse(fut.isDone());
-
-        REGISTER.moveRevision(0L).join();
-
-        assertTrue(fut.isDone());
-
-        assertEquals(TEST_VALUE, fut.get());
-    }
-
-    /**
-     * Test asynchronous update closure.
-     */
-    @Test
-    public void testAsyncUpdate() {
-        VersionedValue<Integer> vv = new VersionedValue<>(REGISTER);
-
-        CompletableFuture<Integer> fut = new CompletableFuture<>();
-
-        vv.update(0L, (v, e) -> fut);
-
-        CompletableFuture<Integer> vvFut = vv.get(0L);
-
-        CompletableFuture<?> revFut = REGISTER.moveRevision(0L);
-
-        assertFalse(fut.isDone());
-        assertFalse(vvFut.isDone());
-        assertFalse(revFut.isDone());
-
-        fut.complete(1);
-
-        revFut.join();
-
-        assertTrue(vvFut.isDone());
-    }
-
-    /**
-     * Test the case when exception happens in updater.
-     */
-    @Test
-    public void testExceptionOnUpdate() {
-        VersionedValue<Integer> vv = new VersionedValue<>(REGISTER, () -> 0);
-
-        final int count = 4;
-        final int successfulCompletionsCount = count / 2;
-
-        AtomicInteger actualSuccessfulCompletionsCount = new AtomicInteger();
-
-        final String exceptionMsg = "test msg";
-
-        for (int i = 0; i < count; i++) {
-            vv.update(0L, (v, e) -> {
-                if (e != null) {
-                    return failedFuture(e);
-                }
-
-                if (v == successfulCompletionsCount) {
-                    throw new IgniteInternalException(exceptionMsg);
-                }
-
-                actualSuccessfulCompletionsCount.incrementAndGet();
-
-                return completedFuture(++v);
-            });
-        }
-
-        AtomicReference<Throwable> exceptionRef = new AtomicReference<>();
-
-        vv.whenComplete((t, v, e) -> exceptionRef.set(e));
-
-        vv.complete(0L);
-
-        assertThrowsWithCause(() -> vv.get(0L).join(), IgniteInternalException.class);
-
-        assertThat(exceptionRef.get().getMessage(), containsString(exceptionMsg));
-    }
-
-    /**
-     * Test with multiple versioned values and asynchronous completion.
+     * Test {@link VersionedValue#whenComplete}.
      */
     @Test
-    public void testAsyncMultiVv() {
-        final String registryName = "Registry";
-        final String assignmentName = "Assignment";
-        final String tableName = "T1_";
-
-        VersionedValue<Map<UUID, String>> tablesVv = new VersionedValue<>(f -> {}, HashMap::new);
-        VersionedValue<Map<UUID, String>> schemasVv = new VersionedValue<>(REGISTER, HashMap::new);
-        VersionedValue<Map<UUID, String>> assignmentsVv = new VersionedValue<>(REGISTER, HashMap::new);
+    public void testWhenComplete() {
```

Comment:
This one seems like a duplicate of the test for incremental versioned value. Is it?

Reply:
If there are many repeated scenarios, I'd recommend extracting them into a base class

Reply:
This is not a duplicate, since two VersionedValue implementations have different API

Reply:
Ok, I thought there was an abstract class

Reply:
There was an abstract class at one point, but it was only used for implementation sharing purposes, not for behavior sharing purposes. Since then I got rid of it and replaced inheritance with composition